### PR TITLE
fix(curriculum): RPG - Step 40 - 41: Passing with wrong code

### LINKED ISSUE #53163: RPG - Step 40 - 41: Passing with wrong code
### Describe the Issue

Edited steps to reflect the new order (was 41-42 before).

The tests will pass if you invoke the callbacks instead of assigning the definition/identifier.

40 passing with incorrect code.
```js
button1.onclick = goStore();
```

41 passing with incorrect code.
```js
button1.onclick = goStore();
button2.onclick = goCave();
button3.onclick = fightDragon();
```

Edit: In fact, as long we match on the start of the function name you can put whatever you want after. So this will pass as well.

```js
button1.onclick = goStoressss;
```

### Affected Page

https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures-v8/learn-basic-javascript-by-building-a-role-playing-game/step-40

https://www.freecodecamp.org/learn/javascript-algorithms-and-data-structures-v8/learn-basic-javascript-by-building-a-role-playing-game/step-41

### Your code

```



```

### Expected behavior

The test should not allow function invocation.

The test should match the actual function identifier.

### Screenshots

_No response_

### System

- Device: [e.g. iPhone 6, Laptop]
- OS: [e.g. iOS 14, Windows 10, Ubuntu 20.04]
- Browser: [e.g. Chrome, Safari]
- Version: [e.g. 22]

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by iprime2**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitPod.

Closes #53163

### CODE REVIEW

**Comment by lasjorg** [comment]
This is not what the issue is about. The tests should not allow for the functions to be invoked/called. Doing so will break the event handlers.

The parentheses `()` should _not_ be allowed when assigning the handler functions.

Incorrect:
```
button1.onclick = goStore();
button2.onclick = goCave();
button3.onclick = fightDragon();
```

Correct:
```
button1.onclick = goStore;
button2.onclick = goCave;
button3.onclick = fightDragon;
```


Also, the element event attribute is `onclick` - all lowercase. You should not change that.

**Comment by iprime2** [comment]
@lasjorg 
image

**Comment by lasjorg** [comment]
Please remove the comments from the code.

I just noticed we do not actually test for the correct function name. As long as we match on the start of the function name you can put whatever you want after. So this will pass as well.

```js
button1.onclick = goStoressss;
```

Also, thinking about this some more, we might want to create new tests specifically for the function invocations.

Maybe someone else will have some feedback on how best to handle this.

**Comment by iprime2** [comment]
I will create a new test to check the name of the function for steps 41 and 42.

**Comment by iprime2** [comment]
@lasjorg Can you please review my pull request.

**Comment by lasjorg** [comment]
We shouldn't require semicolons, as such they can't be used for the function name test.


The assert for the function name should be (replace the name and button as needed)

> You should assign the `goStore` function reference to `button1.onclick`.

The assert for the function invocation should be (replace the name as needed):

> You should not call the `goStore` function. Remove the parentheses.

**Comment by iprime2** [comment]
Hey! So the test should pass without a semicolon; only then should it check the function name.
and have to change this from:
You should assign the `goStore` function reference to `button1.onclick.`
to:
You should not call the `goStore` function. Remove the parentheses.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a3c2fccf186146b59c6e96.md [hunk 1]
```diff
@@ -31,10 +31,16 @@ You should not use `let` or `const`.
 assert.notMatch(code, /(let|const)\s+button1\.onclick/);
 ```
 
-You should assign the `goStore` function reference to `button1.onclick`.
+You should assign correct function name `goStore` reference to `button1.onclick`.
 
 ```js
-assert.match(code, /button1\.onclick\s*=\s*goStore/);
+assert.match(code, /button1\.onclick\s*=\s*goStore;/);
```
Comment (lasjorg) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a3c2fccf186146b59c6e96.md:37]:
```suggestion
assert.match(code, /button1\.onclick\s*=\s*goStore\s*;?\s*$/m);
```
Using the `m` flag should allow us to match on the end.

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a3c4a0e52767482c5202d4.md [hunk 2]
```diff
@@ -25,10 +25,16 @@ You should not use `let` or `const` to assign `button2.onclick`.
 assert.notMatch(code, /(let|const)\s+button2\.onclick/);
 ```
 
-You should set the `onclick` property of `button2` to the variable `goCave`.
+You should assign correct function name `goCave` reference to `button2.onclick`.
 
 ```js
-assert.match(code, /button2\.onclick\s*=\s*goCave/);
+assert.match(code, /button2\.onclick\s*=\s*goCave;/);
```
Comment (lasjorg) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a3c4a0e52767482c5202d4.md:31]:
```suggestion
assert.match(code, /button2\.onclick\s*=\s*goCave\s*;?\s*$/m);
```

File: curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a3c4a0e52767482c5202d4.md [hunk 3]
```diff
@@ -43,10 +49,16 @@ You should not use `let` or `const` to assign `button3.onclick`.
 assert.notMatch(code, /(let|const)\s+button3\.onclick/);
 ```
 
-You should set the `onclick` property of `button3` to the variable `fightDragon`.
+You should assign correct function name `fightDragon` reference to `button3.onclick`.
+
+```js
+assert.match(code, /button3\.onclick\s*=\s*fightDragon;/);
```
Comment (lasjorg) [inline curriculum/challenges/english/15-javascript-algorithms-and-data-structures-22/learn-basic-javascript-by-building-a-role-playing-game/62a3c4a0e52767482c5202d4.md:55]:
```suggestion
assert.match(code, /button3\.onclick\s*=\s*fightDragon\s*;?\s*$/m);
```

**Comment by lasjorg** [comment]
Please make all the requested changes.

If the "Commit suggestion" button is still active that change wasn't made and still needs to be made. You can use the "Commit suggestion" button to add the code as well.

**Review by lasjorg — CHANGES_REQUESTED** [review]
There are some issues and missing changes but never mind that. Read on.

---

The function invocation test will never run as is. The test before it will always catch the invocation.

I'm sorry to do this, but I think I have changed my mind and we should just remove the dedicated invocation tests.

I say we just combine them so we are left with just one test for each handler assignment.

---

Step 40

You should assign the `goStore` function reference to `button1.onclick`. Make sure not to call the function.

```js
assert.match(code, /button1\.onclick\s*=\s*goStore\s*;?\s*$/m);
```

---

Step 41

You should assign the `goStore` function reference to `button1.onclick`. Make sure not to call the function.

```js
assert.match(code, /button1\.onclick\s*=\s*goStore\s*;?\s*$/m);
```

You should assign the `goCave` function reference to `button2.onclick`. Make sure not to call the function.

```js
assert.match(code, /button2\.onclick\s*=\s*goCave\s*;?\s*$/m);
```

You should assign the `fightDragon` function reference to `button3.onclick`. Make sure not to call the function.

```js
assert.match(code, /button3\.onclick\s*=\s*fightDragon\s*;?\s*$/m);
```

Rechecking the `goStore` function in step 41 can't hurt just to make sure they do not change that in the process of adding the other two handlers.

---

Again, sorry for the extra work required but I think this change is for the best.

**Comment by iprime2** [comment]
hey @lasjorg as per your suggestion I implemented `assert.match(code, /button1\.onclick\s*=\s*goStore\s*;?\s*$/);` method to check but the test fails if I provide this correct input `button1.onclick = goStore;` .

**Review by iprime2 — COMMENTED** [review]
Solved.

**Review by lasjorg — CHANGES_REQUESTED** [review]
I think the code review is getting a little messy so instead here are the tests for the two steps in their entirety. This is exactly how they should look. If you have any doubts please just copy and paste them.


## Step 40

````md

# --hints--

You should use dot notation to access the `onclick` property of `button1`.

```js
assert.match(code, /button1\.onclick/);
```

You should not use `let` or `const`.

```js
assert.notMatch(code, /(let|const)\s+button1\.onclick/);
```

You should assign the `goStore` function reference to `button1.onclick`. Make sure not to call the function.

```js
assert.match(code, /button1\.onclick\s*=\s*goStore\s*;?\s*$/m);
```

````

---

## Step 41

````md

# --hints--

You should assign the `goStore` function reference to `button1.onclick`. Make sure not to call the function.

```js
assert.match(code, /button1\.onclick\s*=\s*goStore\s*;?\s*$/m);
```

You should use dot notation to access the `onclick` property of `button2`.

```js
assert.match(code, /button2\.onclick/);
```

You should not use `let` or `const` to assign `button2.onclick`.

```js
assert.notMatch(code, /(let|const)\s+button2\.onclick/);
```

You should assign the `goCave` function reference to `button2.onclick`. Make sure not to call the function.

```js
assert.match(code, /button2\.onclick\s*=\s*goCave\s*;?\s*$/m);
```

You should use dot notation to access the `onclick` property of `button3`.

```js
assert.match(code, /button3\.onclick/);
```

You should not use `let` or `const` to assign `button3.onclick`.

```js
assert.notMatch(code, /(let|const)\s+button3\.onclick/);
```

You should assign the `fightDragon` function reference to `button3.onclick`. Make sure not to call the function.

```js
assert.match(code, /button3\.onclick\s*=\s*fightDragon\s*;?\s*$/m);
```

````

**Review by lasjorg — APPROVED** [review]
This looks good.

Thank you, and sorry for the extra work, but we got there.

**Review by Sembauke — APPROVED** [review]

**Pull request merged**